File: frontend/src/utils/directionCode.ts
/** OnlineTest Direction.name (DI, TPI, PI, …) — fayl/fan nomidan. */

export function matchDirectionCode(value: string | null | undefined, allowed: string[]): string {
  const key = (value || '').trim().toLowerCase();
  if (!key) return '';
  return allowed.find((code) => code.trim().toLowerCase() === key)?.trim() || '';
}

export function inferDirectionCode(text: string, allowed: string[]): string {
  const pool = allowed.map((c) => c.trim()).filter(Boolean);
  if (!text || !pool.length) return '';
  const base = text.replace(/\.(pdf|docx?|xlsx?)$/i, '').trim();
  const paren = base.match(/\(([^)]+)\)\s*$/);
  if (paren?.[1]) {
    const hit = matchDirectionCode(paren[1], pool);
    if (hit) return hit;
  }
  const sorted = [...pool].sort((a, b) => b.length - a.length);
  for (const code of sorted) {
    const escaped = code.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const re = new RegExp(`(?:^|[^\\p{L}\\p{N}])${escaped}(?:[^\\p{L}\\p{N}]|$)`, 'iu');
    if (re.test(base)) return code;
  }
  return '';
}

export function resolveSyllabusDirection(
  row: { direction_code?: string | null; subject_name?: string; file_name?: string },
  allowed: string[],
): string {
  return (
    matchDirectionCode(row.direction_code, allowed) ||
    inferDirectionCode(row.subject_name || '', allowed) ||
    inferDirectionCode(row.file_name || '', allowed)
  );
}

File: backend_fastapi/app/services/direction_code.py
"""OnlineTest Direction.name (DI, TPI, PI, …) ni fayl/fan nomidan aniqlash."""

from __future__ import annotations

import re

# Uzunroq kodlar avval — TPI ichidagi PI, SSBJSS ichidagi S.
DEFAULT_DIRECTION_CODES: tuple[str, ...] = (
    "SSBJSS",
    "TBATM",
    "OHI",
    "TPI",
    "RTT",
    "DI",
    "FT",
    "PI",
    "BM",
    "MD",
    "XT",
    "ЛД",
    "F",
    "S",
    "P",
)


def normalize_direction_code(raw: str | None, allowed: list[str] | None = None) -> str:
    value = (raw or "").strip()[:32]
    if not value:
        return ""
    pool = [c.strip() for c in (allowed or list(DEFAULT_DIRECTION_CODES)) if c and str(c).strip()]
    if not pool:
        return value
    key = value.casefold()
    for code in pool:
        if code.casefold() == key:
            return code
    return value


def infer_direction_code(text: str, allowed: list[str] | None = None) -> str:
    """`(TPI)` yoki so'z chegarasidagi kodni topadi."""
    pool = [c.strip() for c in (allowed or list(DEFAULT_DIRECTION_CODES)) if c and str(c).strip()]
    if not text or not pool:
        return ""
    base = re.sub(r"\.(pdf|docx?|xlsx?)$", "", text, flags=re.IGNORECASE).strip()
    paren = re.search(r"\(([^)]+)\)\s*$", base)
    if paren:
        hit = normalize_direction_code(paren.group(1), pool)
        if hit.casefold() in {c.casefold() for c in pool}:
            return hit
    pool_sorted = sorted(pool, key=lambda c: len(c), reverse=True)
    for code in pool_sorted:
        pattern = rf"(?<![\w]){re.escape(code)}(?![\w])"
        if re.search(pattern, base, flags=re.IGNORECASE):
            return code
    return ""
